Custom Metrics Fixes And Improvements

dustman9000 opened the pull request:
* Fix SQL statement to conform to crt.sh SQL statements (eg. https://crt.sh/?Identity=devshift.org&exclude=expired&deduplicate=Y&showSQL=Y)
* Fix issue where operator-sdk controller-runtime metrics http server port 8080 conflicts with operator-custom-metrics http metrics server port. Verified custom metrics are missing with latest from master on hive-stage.
* Remove unused code
* Add improved error logs
* Defensive check against null private account key to prevent runtime crash

Tested locally with a prometheus instance. Verified proper cert issued counts show up in custom metrics endpoint.

```
$ kubectl run -it --rm debug --image=radial/busyboxplus:curl --restart=Never -- curl http://172.17.0.3:8080/metrics
# HELP certman_operator_certificate_issue_duration Runtime of issue certificate function in seconds
# TYPE certman_operator_certificate_issue_duration summary
certman_operator_certificate_issue_duration_sum{name="certman-operator"} 0
certman_operator_certificate_issue_duration_count{name="certman-operator"} 0
# HELP certman_operator_certs_in_last_day_devshift_org Report how many certs have been issued for Devshift.org in the last 24 hours
# TYPE certman_operator_certs_in_last_day_devshift_org gauge
certman_operator_certs_in_last_day_devshift_org{name="certman-operator"} 223
# HELP certman_operator_certs_in_last_day_openshift_apps_com Report how many certs have been issued for Openshiftapps.com in the last 24 hours
# TYPE certman_operator_certs_in_last_day_openshift_apps_com gauge
certman_operator_certs_in_last_day_openshift_apps_com{name="certman-operator"} 56
# HELP certman_operator_certs_in_last_week_devshift_org Report how many certs have been issued for Devshift.org in the last 7 days
# TYPE certman_operator_certs_in_last_week_devshift_org gauge
certman_operator_certs_in_last_week_devshift_org{name="certman-operator"} 1413
# HELP certman_operator_certs_in_last_week_openshift_apps_com Report how many certs have been issued for Openshiftapps.com in the last 7 days
# TYPE certman_operator_certs_in_last_week_openshift_apps_com gauge
certman_operator_certs_in_last_week_openshift_apps_com{name="certman-operator"} 371
...
```

dustman9000 commented on pkg/localmetrics/crt_sh.go:
```diff
@@ -147,5 +38,5 @@ func GetCountOfCertsIssued(domain string, durationDays int) int {
 
 func getPsqlInfo() string {
-	return fmt.Sprintf("host=%s port=%d user=%s dbname=%s sslmode=disable", CRT_SH_PG_DB_HOSTNAME, CRT_SH_PG_DB_PORT, CRT_SH_PG_DB_USERNAME, CRT_SH_PG_DB_NAME)
+	return fmt.Sprintf("host=%s port=%d user=%s dbname=%s sslmode=disable binary_parameters=yes", CRT_SH_PG_DB_HOSTNAME, CRT_SH_PG_DB_PORT, CRT_SH_PG_DB_USERNAME, CRT_SH_PG_DB_NAME)
```
Fix for PSQL error: `unnamed prepared statement does not exist`

fahlmant commented on cmd/manager/main.go:
```diff
@@ -149,16 +147,19 @@ func main() {
-	if detectSDKLocal == "local" {
-		log.Info("Skipping metrics configuration; not running in a cluster.")
+	// Get the namespace the operator is currently deployed in.
+	if _, err := k8sutil.GetOperatorNamespace(); err != nil {
```
Just out of curiosity, why change this from the current pattern?

dustman9000 replied:
I was having issues getting the metrics-server running locally and these changes were remnants of that troubleshooting. This is mainly to bring it up to date with operator-sdk (https://github.com/operator-framework/operator-sdk/blob/master/website/content/en/docs/upgrading-sdk-version/version-upgrade-guide.md).

fahlmant replied:
:+1:

fahlmant commented on cmd/manager/main.go:
```diff
@@ -149,16 +147,19 @@ func main() {
 		if err := metrics.ConfigureMetrics(context.TODO(), *metricsServer); err != nil {
 			log.Error(err, "Failed to configure Metrics")
 			os.Exit(1)
+		} else {
```
I think this could just be
```
if err := metrics.ConfigureMetrics(context.TODO(), *metricsServer); err != nil {
        log.Error(err, "Failed to configure Metrics")
        os.Exit(1)
}
log.Info("Successfully configured Metrics")
```
Just saves a line of code

dustman9000 commented:
/assign @yithian

dustman9000 commented on cmd/manager/main.go:
```diff
@@ -35,15 +36,12 @@ import (
 	metricsHost                   = "0.0.0.0"
 	metricsPath                   = "/metrics"
-	metricsPort                   = "8080"
+	metricsPort                   = "8081"
```
To de-conflict with controller-runtime default port/path.

2uasimojo commented on cmd/manager/main.go:
```diff
@@ -95,6 +95,8 @@ func main() {
 	// Set default manager options
 	options := manager.Options{
 		Namespace: namespace,
+		// Disable controller-runtime metrics serving
+		MetricsBindAddress: "0",
```
++

yithian commented on cmd/manager/main.go:
```diff
@@ -149,17 +149,19 @@ func main() {
 			log.Error(err, "Failed to configure Metrics")
 			os.Exit(1)
 		}
+		log.Info("Successfully configured Metrics")
```
I assume this is just for clarity? I don't see anything obviously changing metrics configuration...

dustman9000 replied:
Yes, this is for clarity.

yithian commented on pkg/controller/certificaterequest/issue_certificate.go:
```diff
@@ -79,10 +80,12 @@ func (r *ReconcileCertificateRequest) IssueCertificate(reqLogger logr.Logger, cr
 
 	err = leClient.CreateOrder(cr.Spec.DnsNames)
 	if err != nil {
+		reqLogger.Error(err, "failed to to creat order")
```
```suggestion
		reqLogger.Error(err, "failed to create order")
```

yithian commented:
What's with the change of metrics from openshift.com to devshift.org?

dustman9000 commented:
> What's with the change of metrics from openshift.com to devshift.org?

We do not request LE certificates for openshift.com. I've removed this metric and replaced with the devshift.org instead.

yithian commented:
/lgtm

@cblecker i wouldn't mind a second set of eyes on the `queries.go` stuff

cblecker commented:
/hold

cblecker commented:
/lgtm
/hold cancel
